File: client/src/models/Operation.ts
import * as monaco from 'monaco-editor/esm/vs/editor/editor.api';

export interface Operation {
  text: string,
  endColumn: number,
  startColumn: number;
  endLine: number;
  startLine: number;
  revisionId: number;
  originatorId?: string;
}

export function fromEvent(event: monaco.editor.IModelContentChangedEvent, originatorId: string, revisionId: number = -1): Operation {
  return {
    revisionId: revisionId,
    text: event.changes[0].text,
    endColumn: event.changes[0].range.endColumn,
    startColumn: event.changes[0].range.startColumn,
    endLine: event.changes[0].range.endLineNumber,
    startLine: event.changes[0].range.startLineNumber,
    originatorId: originatorId,
  };
}

export function copy(op: Operation): Operation {
  return {
    revisionId: op.revisionId,
    text: op.text,
    endColumn: op.endColumn,
    startColumn: op.startColumn,
    startLine: op.startLine,
    endLine: op.endLine,
    originatorId: op.originatorId,
  };
}

export type OpHistory = { [revId: number]: Operation };

File: server/src/models/Operation.cs
namespace server.models
{
    public class Operation
    {
        public string Text { get; set; }
        public int StartColumn { get; set; }
        public int EndColumn { get; set; }
        public int StartLine { get; set; }
        public int EndLine { get; set; }
        public int RevisionId { get; set; }
        public string OriginatorId { get; set; }
        public string Id { get; set; } = Guid.NewGuid().ToString();
        public bool IsSimpleInsert { 
            get 
            {
                return StartColumn == EndColumn && StartLine == EndLine;
            }
        }

        public new string ToString
        {
            get
            {
                return $"{StartColumn}, {EndColumn}, {StartLine}, {EndLine} | {RevisionId} | {Text}";
            }
        }

        public Operation DeepCopy()
        {
            return new Operation
            {
                Id = this.Id,
                Text = this.Text,
                StartColumn = this.StartColumn,
                EndColumn = this.EndColumn,
                StartLine = this.StartLine,
                EndLine = this.EndLine,
                RevisionId = this.RevisionId,
                OriginatorId = this.OriginatorId
            };
        }
    }   
}